sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala

```diff
@@ -389,6 +394,16 @@ class V2ExpressionBuilder(e: Expression, isPredicate: Boolean = false) {
     }
   }
 
+  private def generatePredicateWithName(
```

Comment:
one idea: can we update `generateExpressionWithName` to take the target expression instead of the children? Then we can simply do
```
val children = e.children
...
if (d.dataType == BooleanType && isPredicate) {
  new V2Predicate ...
} else {
  new GeneralScalarExpression
}
```

Reply:
sounds good, I will try that.